src/platform_impl/web/event_loop/proxy.rs

```diff
@@ -1,26 +1,96 @@
-use super::runner;
-use crate::event::Event;
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::mpsc::{self, Receiver, RecvError, SendError, Sender, TryRecvError};
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
 use crate::event_loop::EventLoopClosed;
 
 pub struct EventLoopProxy<T: 'static> {
-    runner: runner::Shared<T>,
+    sender: AsyncSender<T>,
 }
 
 impl<T: 'static> EventLoopProxy<T> {
-    pub fn new(runner: runner::Shared<T>) -> Self {
-        Self { runner }
+    pub fn new(sender: AsyncSender<T>) -> Self {
+        Self { sender }
     }
 
     pub fn send_event(&self, event: T) -> Result<(), EventLoopClosed<T>> {
-        self.runner.send_event(Event::UserEvent(event));
-        Ok(())
+        match self.sender.send(event) {
+            Ok(()) => Ok(()),
+            Err(SendError(val)) => Err(EventLoopClosed(val)),
+        }
     }
 }
 
 impl<T: 'static> Clone for EventLoopProxy<T> {
     fn clone(&self) -> Self {
         Self {
-            runner: self.runner.clone(),
+            sender: self.sender.clone(),
+        }
+    }
+}
+
+pub fn channel<T: 'static>() -> (AsyncSender<T>, AsyncReceiver<T>) {
+    let (sender, receiver) = mpsc::channel();
+    let waker = Arc::new(Mutex::new(None));
+
+    let sender = AsyncSender {
+        sender,
+        waker: Arc::clone(&waker),
+    };
+    let receiver = AsyncReceiver { receiver, waker };
+
+    (sender, receiver)
+}
+
+pub struct AsyncSender<T: 'static> {
+    sender: Sender<T>,
+    waker: Arc<Mutex<Option<Waker>>>,
```

Comment:
Is `Mutex` available on WASM? I always forget what synchronization primitives panic on block for non-atomic WASM or not. Although, it looks like that, in single threaded cases, it shouldn't block. Is this right?

Reply:
> Is `Mutex` available on WASM?

Yes.

> Although, it looks like that, in single threaded cases, it shouldn't block. Is this right?

For single threaded cases it shouldn't block indeed, but it can for multi-threaded cases. Considering we are just storing a single value and not holding a lock in practice it's not a problem.

I would still like to propose that we don't use a `Mutex` here and use [`AtomicWaker`](https://docs.rs/atomic-waker/1.1.0/atomic_waker/struct.AtomicWaker.html) instead. The crate has [no dependencies](https://crates.io/crates/atomic-waker/1.1.0/dependencies) and is [really small](https://docs.rs/atomic-waker/1.1.0/src/atomic_waker/lib.rs.html).

If the dependency is still a concern, I would like to point to my suggestion above: hiding the `Send` support behind the `+atomics` target feature.

Reply:
I would support using `atomic-waker`, but I'd ask the rest of the maintainers first.